File: merge-two-sorted-lists.go
package main

import(
    "fmt"
)

type ListNode struct {
    Val int
    Next *ListNode
}

func mergeTwoLists(l1 *ListNode, l2 *ListNode) *ListNode {
    if l1 == nil {
        return l2
    }
    if l2 == nil {
        return l1
    }

    root := &ListNode{}
    c := root
    for {
        if l1 != nil && l2 != nil {
            if l1.Val == l2.Val {
                t1, t2 := l1, l2
                l1, l2 = l1.Next, l2.Next
                c.Next, c = t1, t1
                c.Next, c = t2, t2
            }else if l1.Val < l2.Val {
                c.Next, c = l1, l1
                l1 = l1.Next
            }else {
                c.Next, c = l2, l2
                l2 = l2.Next
            }
        }else if l1 == nil {
            c.Next = l2
            break
        }else {
            c.Next = l1
            break
        }
    }
    return root.Next
}

func gen(l []int) *ListNode {
    root := &ListNode{}
    c := root
    for _, v := range l {
        n := &ListNode{Val: v}
        c.Next = n
        c = c.Next
    }
    return root.Next
}

func printList(l *ListNode) {
    if l == nil {
        fmt.Println("")
    }else {
        for l != nil {
            fmt.Printf("%d->", l.Val)
            l = l.Next
        }
        fmt.Println()
    }
}

func main() {
    l1 := gen([]int {2,2,4,5,7,8})
    l2 := gen([]int {1,3,4,5})

    l3 := mergeTwoLists(l1, l2)
    printList(l3)
}

File: merge-two-sorted-lists.py
#! /usr/bin/env python
# _*_ coding:utf-8 _*_


class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):
    def mergeTwoLists(self, l1, l2):
        if not l1:
            return l2
        if not l2:
            return l1

        root = ListNode(-1)
        c = root
        while True:
            if l1 and l2:
                if l1.val == l2.val:
                    t1, t2 = l1, l2
                    l1, l2 = l1.next, l2.next
                    c.next, c = t1, t1
                    c.next, c = t2, t2
                elif l1.val < l2.val:
                    t,l1 = l1, l1.next
                    c.next, c = t, t
                elif l1.val > l2.val:
                    t, l2 = l2, l2.next
                    c.next, c = t, t
            elif not l1:
                c.next = l2
                break
            elif not l2:
                c.next = l1
                break
        return root.next


def genList(l):
    l.sort()
    root = ListNode(0)
    c = root
    for x in l:
        n = ListNode(x)
        c.next = n
        c = n
    return root.next


def printList(n):
    if not n:
        print ""
        return
    s = ''
    while n:
        s += str(n.val) + '->'
        n = n.next
    print s[:-2]


if __name__ == '__main__':
    so = Solution()
    l1 = genList([1,2,4])
    l2 = genList([1,3,4])
    printList(so.mergeTwoLists(l1, l2))
    l1 = genList([1, 4, 9, 11, 13])
    l2 = genList([1, 3, 4])
    printList(so.mergeTwoLists(l1, l2))
    l1 = genList([2,4,9,11])
    l2 = genList([1,3,4])
    printList(so.mergeTwoLists(l1, l2))
    l1 = genList([])
    l2 = genList([])
    printList(so.mergeTwoLists(l1, l2))
    l1 = genList([])
    l2 = genList([1,3,4])
    printList(so.mergeTwoLists(l1, l2))
    l1 = genList([2,3,4])
    l2 = genList([6,7,8,9,10])
    printList(so.mergeTwoLists(l1, l2))
